adding more integration tests to grid3_client_ts using jest

mohamedamer453 opened the pull request:
### Description

Adding more integration tests to grid3_client_ts using jest.

### Modules covered in the added tests
- Vm.
- K8s.
- Qsfs.
- Zdb.
- KVStore.
- Gateways.
- Capacity Planner.

AhmedHanafy725 reviewed (changes requested):
- Please make sure that the tests are passing on the CI.
- Can we ensure that the deployed machines and k8s are running by ssh to them?
- For the gateway test, can we also deploy a VM with a running server and make the domain point to this server?
- For zdbs, can we get a client for this deployed namespace and make sure we can do essential operations on them?

mohamedamer453 commented on tests/modules/capacity_planner.test.ts:
```diff
@@ -0,0 +1,173 @@
+    const farms: FarmsGetModel = {
+        page: page,
+        maxResult: maxResult,
+    };
```
all of these methods were removed and the models are now used directly in each test.

mohamedamer453 commented on tests/modules/capacity_planner.test.ts:
```diff
@@ -0,0 +1,173 @@
+
+    const res = await gridClient.capacity.filterNodes(filter);
+    log(res);
+    expect(res[0].total_resources["mru"] - res[0].used_resources["mru"]).toBeGreaterThanOrEqual(bytesToGB(mru));
```
The logic of this test was improved and a loop was added to iterate over each of the returned nodes and verify its resources.

mohamedamer453 commented:
### What was added in the last commit :

- Added an ssh function as requested to ssh to VMs/K8s and verify resources, env vars, etc...
- Removed the methods that had the models in them and the models are now used directly in each test.
- Added test steps/scenario for each test.
- Added Error handling for the ssh function to avoid any hanging connections in case of test failure.
- Edited the used flist in the test to the latest one.
- Added more logic to the workflow to install Yggdrasil and generate an ed25519 key-pair.

### what is left in these tests:

- K8s waits Improvement, currently the waits are not dynamic and you have to wait for the master/workers to be ready (still trying to figure out the best way to do it)
- Gateway test, Verifying that the VM was exposed successfully over the Gateway still needs to be done.
- ZDB, using redis-client to verify the namespace and doing essential zdb operations still needs to be done.

### Issues i'm still trying to figure out

- Sometimes the Multiple VMs test fails due to the VMs list randomly changing the position of the deployed VMs in the list, for example sometimes the first VM gets an index of [0] and sometimes it's [1]
- Sometimes the Kubernetes With QSFS test fails due to the error mentioned [here](https://github.com/threefoldtech/grid3_client_ts/issues/412) i'm still trying to figure out if i'm doing the capacity filtering for this test incorrectly or if it's due to the mentioned issue.

A-Harby commented on tests/modules/vm.test.ts:
```diff
@@ -10,65 +10,536 @@ beforeAll(async () => {
+        farmId: 1,
         availableFor: await gridClient.twins.get_my_twin_id(),
     } as FilterOptions);
     const nodeId = +randomChoice(nodes).nodeId;
```
Maybe Node Selection should handles the situation where there is no nodes available (so it could skip the test or re-randomized test data to lower values).

mohamedamer453 replied:
added in https://github.com/threefoldtech/grid3_client_ts/pull/405/commits/7261b52afc18478462fbe1a1c2fbe5af2f893b16 

Error handling for node selection was added and in case a node was not found with the generated resources the error will be caught and new test data will be generated with lower resources, with the maximum value being the old resource value.

A-Harby commented on tests/modules/vm.test.ts:
```diff
@@ -10,65 +10,536 @@ beforeAll(async () => {
+
+    const vm1NodeId = +randomChoice(vm1Node).nodeId;
+    let vm2NodeId = +randomChoice(vm2Node).nodeId;
+    if (vm1NodeId == vm2NodeId) {
```
Shouldn't it be a while instead of if

A-Harby commented:
- Add assertion that node id created is the same as node id given.
- If it's possible, could also add assertion when deleting the contracts, to make sure it's successfully deleted.

mohamedamer453 commented:
> * Add assertion that node id created is the same as node id given.
> * If it's possible, could also add assertion when deleting the contracts, to make sure it's successfully deleted.

Added in https://github.com/threefoldtech/grid3_client_ts/pull/405/commits/0836f31df981a69e0c4fa082f3e47bb7bdd3553b

and also added docs for configuring/running the tests.

AhmedHanafy725 reviewed (changes requested):
why is the report.xml in the PR?

AhmedHanafy725 commented on package.json:
```diff
@@ -25,6 +25,7 @@
         "http-server": "^14.0.0",
         "libsodium-wrappers": "^0.7.10",
         "netaddr": "^1.1.0",
+        "node-ssh": "^13.0.0",
```
please add this package to the dev dependencies
